Firmware/RTK_Surveyor/Begin.ino

```diff
@@ -273,80 +274,111 @@ void beginSD()
-      settings.spiFrequency = 16;
-    }
+      //If an SD card is present, allow SdFat to take over
+      log_d("SD card detected - using SPI and SdFat");
 
-    resetSPI(); //Re-initialize the SPI/SD interface
-
-    if (sd->begin(SdSpiConfig(pin_microSD_CS, SHARED_SPI, SD_SCK_MHZ(settings.spiFrequency))) == false)
-    {
-      tries = 0;
-      maxTries = 1;
-      for ( ; tries < maxTries ; tries++)
+      //Allocate the data structure that manages the microSD card
+      if (USE_SPI_MICROSD && !sd)
       {
-        log_d("SD init failed. Trying again %d out of %d", tries + 1, maxTries);
-
-        delay(250); //Give SD more time to power up, then try again
-        if (sd->begin(SdSpiConfig(pin_microSD_CS, SHARED_SPI, SD_SCK_MHZ(settings.spiFrequency))) == true) break;
+        sd = new SdFat();
+        if (!sd)
+        {
+          log_d("Failed to allocate the SdFat structure!");
+          break;
+        }
       }
-
-      if (tries == maxTries)
+  
+      if (settings.spiFrequency > 16)
       {
-        systemPrintln("SD init failed. Is card formatted?");
-        digitalWrite(pin_microSD_CS, HIGH); //Be sure SD is deselected
-
-        //Check reset count and prevent rolling reboot
-        if (settings.resetCount < 5)
+        systemPrintln("Error: SPI Frequency out of range. Default to 16MHz");
+        settings.spiFrequency = 16;
+      }
+  
+      resetSPI(); //Re-initialize the SPI/SD interface
+  
+      if (sd->begin(SdSpiConfig(pin_microSD_CS, SHARED_SPI, SD_SCK_MHZ(settings.spiFrequency))) == false)
+      {
+        tries = 0;
+        maxTries = 1;
+        for ( ; tries < maxTries ; tries++)
+        {
+          log_d("SD init failed - using SPI and SdFat. Trying again %d out of %d", tries + 1, maxTries);
+  
+          delay(250); //Give SD more time to power up, then try again
+          if (sd->begin(SdSpiConfig(pin_microSD_CS, SHARED_SPI, SD_SCK_MHZ(settings.spiFrequency))) == true) break;
+        }
+  
+        if (tries == maxTries)
         {
-          if (settings.forceResetOnSDFail == true)
-            ESP.restart();
+          systemPrintln("SD init failed - using SPI and SdFat. Is card formatted?");
+          digitalWrite(pin_microSD_CS, HIGH); //Be sure SD is deselected
+  
+          //Check reset count and prevent rolling reboot
+          if (settings.resetCount < 5)
+          {
+            if (settings.forceResetOnSDFail == true)
+              ESP.restart();
+          }
+          break;
         }
+      }
+  
+      //Change to root directory. All new file creation will be in root.
+      if (sd->chdir() == false)
+      {
+        systemPrintln("SD change directory failed");
         break;
       }
     }
-
-    //Change to root directory. All new file creation will be in root.
-    if (sd->chdir() == false)
+    else
     {
-      systemPrintln("SD change directory failed");
-      break;
+      // SDIO MMC
+      if (SD_MMC.begin() == false)
```

Comment:
We use beginSD() and the sdPresent() function as a software card-detect (in absence of hardware CD). I believe you have hardware CD so please be sure the system fails gracefully if an SD is not present. Also please test that hot insert/removal is supported.